### ISSUE TITLE
C source files are not recompiled when cc-options change

### ISSUE THREAD

**Issue by tmcdonell** [haskell/cabal#4937]
My project has some C files, whose behaviour depends on whether we are compiling the project in debug mode or not. This is accomplished via a `cc-options: -DACCELERATE_DEBUG` in the .cabal file ([see here](https://github.com/AccelerateHS/accelerate/blob/40adaddc1fd7f137f62b99dcb617f5d32671b16c/accelerate.cabal#L416)). However, switching between regular and debug builds does not recompile the C sources (though it does seem to recompile relevant .hs files).

Related: https://github.com/haskell/cabal/issues/4306

**Comment by ezyang** [comment]
Yes, the recomp avoidance logic for C files is just... bad. Improvements would be much appreciated.
